File: config/cloud_init.py
import yaml
import json
import uuid
from dataclasses import dataclass, field


@dataclass
class UserData:
    hostname: str
    ssh_user: str
    ssh_public_keys_content: list[str]

    def to_ci_format(self) -> str:
        user_data_content = {
            "hostname": self.hostname,
            "manage_etc_hosts": True,
            "disable_root_pw": True,
            "users": [
                {
                    "name": self.ssh_user,
                    "sudo": "ALL=(ALL) NOPASSWD:ALL",
                    "ssh_authorized_keys": self.ssh_public_keys_content,
                }
            ],
        }

        return (
            "#cloud-config"
            + "\n"
            + yaml.dump(
                user_data_content, indent=2, default_flow_style=False, sort_keys=False
            )
        )


@dataclass
class NetworkConfig:
    ip_address: str
    nameservers: list[str]
    gateway: str
    mac_address: str

    def to_ci_format(self) -> str:
        network_config = {
            "network": {
                "version": 2,
                "ethernets": {
                    "eth0": {
                        "dhcp4": False,
                        "addresses": [f"{self.ip_address}/24"],
                        "gateway4": self.gateway,
                        "macaddress": self.mac_address,
                        "nameservers": {
                            "addresses": self.nameservers,
                        },
                    }
                },
            },
        }

        return yaml.dump(
            network_config, indent=2, default_flow_style=False, sort_keys=False
        )


@dataclass
class MetaData:
    hostname: str
    prefix: str = "monochromatic"
    instance_id: str = field(init=False)

    def __post_init__(self):
        self.instance_id = f"{self.prefix}{self.hostname}-{uuid.uuid4().hex[:8]}"

    def to_ci_format(self) -> str:
        meta_data = {
            "instance-id": self.instance_id,
            "local-hostname": self.hostname,
        }
        return json.dumps(meta_data, indent=2)


class CloudInitConfig:
    user_data: UserData
    meta_data: MetaData
    network_config: NetworkConfig

    # TODO: add required fields in the arguments
    def __init__(
        self,
        hostname: str,
        ssh_user: str,
        ssh_public_keys: list[str],
        ipv4_address: str,
        nameservers: list[str],
        ipv4_gateway_address: str,
        mac_address: str,
    ):
        self.user_data = UserData(
            hostname=hostname,
            ssh_user=ssh_user,
            ssh_public_keys_content=ssh_public_keys,
        )

        self.meta_data = MetaData(hostname=hostname)

        self.network_config = NetworkConfig(
            ip_address=ipv4_address,
            nameservers=nameservers,
            gateway=ipv4_gateway_address,
            mac_address=mac_address,
        )

    @staticmethod
    def from_dict(data: dict):
        return CloudInitConfig(
            hostname=data["hostname"],
            ssh_user=data["ssh_user"],
            ssh_public_keys=data["ssh_publickeys"],
            ipv4_address=data["ipv4_address"],
            nameservers=data["nameservers"],
            ipv4_gateway_address=data["ipv4_gateway_address"],
            mac_address=data["mac_address"],
        )

File: go/internal/config/cloud_init.go
package config

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/google/uuid"
	"github.com/nnurry/harmonia/internal/utils"
)

type userData struct {
	Hostname              string
	SshUser               string
	PublicSshKeysContents []string
	DisableRootPW         bool
}

type metaData struct {
	Hostname   string
	InstanceId string `json:"instance-id"`
}

type networkConfig struct {
	IpV4Addresses    []string
	Nameservers      []string
	GatewayV4Address string
	MacAddress       string
	Dhcp4            bool
}

type cloudInitUser struct {
	Name              string   `yaml:"name"`
	Sudo              string   `yaml:"sudo,omitempty"`
	SSHAuthorizedKeys []string `yaml:"ssh_authorized_keys"`
}

type cloudConfigUserDataYAML struct {
	Hostname       string          `yaml:"hostname"`
	ManageEtcHosts bool            `yaml:"manage_etc_hosts"`
	DisableRootPW  bool            `yaml:"disable_root_pw"`
	Users          []cloudInitUser `yaml:"users"`
}

type networkConfigNameserversYAML struct {
	Addresses []string `yaml:"addresses"`
}

type networkConfigEthernetYAML struct {
	Dhcp4       bool                         `yaml:"dhcp4"`
	Addresses   []string                     `yaml:"addresses,omitempty"`
	Gateway4    string                       `yaml:"gateway4,omitempty"`
	MacAddress  string                       `yaml:"macaddress,omitempty"`
	Nameservers networkConfigNameserversYAML `yaml:"nameservers,omitempty"`
}

type networkConfigNetworkYAML struct {
	Version   int                                  `yaml:"version"`
	Ethernets map[string]networkConfigEthernetYAML `yaml:"ethernets"`
}

type cloudConfigNetworkConfigYAML struct {
	Network networkConfigNetworkYAML `yaml:"network"`
}

type CloudInitConfig struct {
	userData      userData
	metaData      metaData
	networkConfig networkConfig
}

func NewCloudInitConfig(
	hostname string,
	sshUser string,
	sshPublicKeys []string,
	ipv4Address string,
	nameservers []string,
	ipv4GatewayAddress string,
	macAddress string,
	disableRootPW bool,
) *CloudInitConfig {
	ci := &CloudInitConfig{}

	ci.userData = userData{
		Hostname:              hostname,
		SshUser:               sshUser,
		PublicSshKeysContents: sshPublicKeys,
		DisableRootPW:         disableRootPW, // Set this value from input
	}

	instanceID := fmt.Sprintf("monochromatic%s-%s", hostname, uuid.New().String()[:8])
	ci.metaData = metaData{
		Hostname:   hostname,
		InstanceId: instanceID,
	}

	ci.networkConfig = networkConfig{
		IpV4Addresses:    []string{fmt.Sprintf("%s/24", ipv4Address)},
		Nameservers:      nameservers,
		GatewayV4Address: ipv4GatewayAddress,
		MacAddress:       macAddress,
		Dhcp4:            false,
	}

	return ci
}

func (ci *CloudInitConfig) ToUserDataYAML() (string, error) {
	user := cloudInitUser{
		Name:              ci.userData.SshUser,
		Sudo:              "ALL=(ALL) NOPASSWD:ALL",
		SSHAuthorizedKeys: ci.userData.PublicSshKeysContents,
	}

	ciYaml := cloudConfigUserDataYAML{
		Hostname:       ci.userData.Hostname,
		ManageEtcHosts: true,
		DisableRootPW:  ci.userData.DisableRootPW,
		Users:          []cloudInitUser{user},
	}

	return utils.MarshalYAML(&ciYaml, "#cloud-config")
}

func (ci *CloudInitConfig) ToMetaDataJSON() (string, error) {
	byteSlice, err := json.MarshalIndent(ci.metaData, "", "  ")
	if err != nil {
		return "", fmt.Errorf("failed to marshal meta data to JSON: %w", err)
	}
	return string(byteSlice), nil
}

func (ci *CloudInitConfig) ToNetworkConfigYAML() (string, error) {
	ethernet := networkConfigEthernetYAML{
		Dhcp4:      ci.networkConfig.Dhcp4,
		MacAddress: ci.networkConfig.MacAddress,
	}

	if !ci.networkConfig.Dhcp4 {
		ethernet.Addresses = ci.networkConfig.IpV4Addresses
		ethernet.Gateway4 = ci.networkConfig.GatewayV4Address
	}

	if len(ci.networkConfig.Nameservers) > 0 {
		ethernet.Nameservers = networkConfigNameserversYAML{
			Addresses: ci.networkConfig.Nameservers,
		}
	}

	networkConfigData := cloudConfigNetworkConfigYAML{
		Network: networkConfigNetworkYAML{
			Version: 2,
			Ethernets: map[string]networkConfigEthernetYAML{
				"eth0": ethernet,
			},
		},
	}

	return utils.MarshalYAML(&networkConfigData, "") // No header for network-config
}

func NewCloudInitFromNodeConfig(node VMNode, group SharedConfigs) (*CloudInitConfig, error) {
	publicKeyPath, err := group.SSH.GetExpandedPublicKeyPath()
	if err != nil {
		return nil, fmt.Errorf("error expanding public key path: %w", err)
	}
	publicKeyContent, err := os.ReadFile(publicKeyPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read public key file %s: %w", publicKeyPath, err)
	}

	return NewCloudInitConfig(
		node.Name,
		group.SSH.User,
		[]string{string(publicKeyContent)},
		node.IPAddress,
		group.CloudInit.Nameservers,
		node.GatewayAddress,
		node.MACAddress,
		group.CloudInit.DisableRootPW,
	), nil
}
